lib/llvmopencl/WorkitemLoops.cc

```diff
@@ -893,6 +891,101 @@ llvm::AllocaInst *WorkitemLoopsImpl::getContextArray(llvm::Instruction *Inst,
 }
 
+/// Tries to rematerialize the given value-defining instruction.
+///
+/// Rematerialization in this context means recomputing the value produced
+/// in the use site instead of storing and loading a once-computed variable
+/// from the context.
+///
+/// \param Before the instruction before which the cloned instructions should
+/// be added.
+/// \param Def is the produced value to attempt to clone recursively.
+/// \param CanDoIt can be set to a true-initialized boolean in which case the
+/// cloning is not actually done, but only its possibility is investigated.
+/// \param Depth the recursion depth. Used to limit rematerialization size.
+/// \return The rematerialized instruction if possible and beneficial.
+llvm::Value *tryToRematerialize(llvm::Instruction *Before, llvm::Value *Def,
+                                bool *CanDoIt = nullptr, int *Depth = 0) {
+
+  // Limit the height of the cloned instruction tree to avoid counter-productive
+  // rematerialization.
+  if (Depth != nullptr && *Depth > 5) {
+    if (CanDoIt != nullptr)
+      *CanDoIt = false;
+#ifdef DEBUG_WORK_ITEM_LOOPS
+    std::cerr << "##### cannot remat (too deep)" << std::endl;
+    Def->dump();
+#endif
+    return nullptr;
+  }
+
+  // A call without arguments: Setup a pre-check before cloning to see if we
+  // can succeed.
+  if (CanDoIt == nullptr && Depth == nullptr) {
+    bool Able = true;
+    int Depth = 0;
+    tryToRematerialize(Before, Def, &Able, &Depth);
+    if (!Able)
+      return nullptr;
+    Depth = 0;
+    return tryToRematerialize(Before, Def, nullptr, &Depth);
+  }
+
+  if (llvm::CallInst *Call = dyn_cast<CallInst>(Def)) {
+    auto Callee = Call->getCalledFunction();
+    if (Callee == nullptr || (Callee->getName() != GID_BUILTIN_NAME &&
+                              Callee->getName() != GS_BUILTIN_NAME &&
+                              Callee->getName() != GROUP_ID_BUILTIN_NAME &&
+                              Callee->getName() != LID_BUILTIN_NAME &&
+                              Callee->getName() != LS_BUILTIN_NAME)) {
+#ifdef DEBUG_WORK_ITEM_LOOPS
+      std::cerr << "##### cannot remat a call" << std::endl;
+      Call->dump();
+#endif
+      if (CanDoIt != nullptr)
+        *CanDoIt = false;
+      return nullptr;
+    }
+  } else if (isa<Constant>(Def) || isa<Argument>(Def)) {
+    if (CanDoIt != nullptr)
+      return nullptr;
+    // No need to clone a constant or function argument, we can refer to the
+    // original directly.
+    return Def;
+  } else if (isa<AllocaInst>(Def)) {
+    if (CanDoIt != nullptr)
+      *CanDoIt = false;
+    return nullptr;
+  }
+
+  if (Depth != nullptr)
+    (*Depth)++;
+
+  llvm::Instruction *Inst = dyn_cast<Instruction>(Def);
+  if (Inst == nullptr) {
+#ifdef DEBUG_WORK_ITEM_LOOPS
+    std::cerr << "##### cannot remat (unsupported value type)" << std::endl;
+    Def->dump();
+#endif
+    if (CanDoIt != nullptr)
+      *CanDoIt = false;
+    return nullptr;
+  }
```

Comment:
Probably should return nullptr if the instruction may write to memory or have side-effects.